icu4c/source/test/intltest/messageformat2test.cpp

```diff
@@ -0,0 +1,416 @@
+// © 2016 and later: Unicode, Inc. and others.
+
+#include "unicode/utypes.h"
+
+#if !UCONFIG_NO_FORMATTING
+
+#include "unicode/messageformat2.h"
+#include "messageformat2test.h"
+
+/*
+Tests reflect the syntax specified in
+
+  https://github.com/unicode-org/message-format-wg/commits/main/spec/message.abnf
+
+as of the following commit from 2023-05-09:
+  https://github.com/unicode-org/message-format-wg/commit/194f6efcec5bf396df36a19bd6fa78d1fa2e0867
+
+   Since the MessageFormat2 "parser" is currently only a validator,
+   the following tests only verify that valid messages are validated by the parser
+   and that invalid messages are rejected with an error reflecting the correct line number
+   and offset for the unexpected character.
+*/
+
+// These tests are from the icu4j test suite under icu4j/main/tests/core/src/com/ibm/icu/dev/test/message2/
+UnicodeString validTestCases[] = {
+    // From Mf2IcuTest.java
+    "{There are {$count} files on {$where}}",
```

Comment:
is it possible to share test cases by using text files? one of MF's deliverables is a "conformance test suite"... extracting these would be much better than hardcoding the cases?

Reply:
I was following what [the previous MessageFormat implementation](https://github.com/unicode-org/icu/blob/main/icu4c/source/test/intltest/tmsgfmt.cpp) does, which is to hardcode the test cases in the file, but I wasn't assuming that was the best solution in the longer term, either. If moving these test cases into a file is a requirement for merging this PR, then I can certainly try to.

Reply:
It's probably easier and better to model after the ICU4J tech preview impl. I def try to make my ICU4C and ICU4J unit tests match up when adding a feature.

Side note: we're trying to create a more comprehensive conformance test suite for i18n functionality here, so this would be another place to put that test suite, in the longer term:
https://github.com/unicode-org/conformance

Reply:
The ICU4J impl does embed the test cases in the test code (for example, https://github.com/unicode-org/icu/blob/main/icu4j/main/tests/core/src/com/ibm/icu/dev/test/message2/MessageFormat2Test.java ), so I assume @echeran's comment means  it's okay to keep the current approach for now, though of course the eventual goal is to move these tests into the conformance test suite once a conformance test suite for MessageFormat 2.0 is created.

Currently, the tests in this PR don't include all of the tests in the ICU4J impl, since I was looking to extract a subset that provides reasonable test coverage for parsing, without adding tests that would increase coverage for the (not-yet-existent) semantic checker but not the parser.